src/index.ts

```diff
@@ -10,25 +10,28 @@ import {pathToIndices} from "./utils";
  * @param entropy optional additional entropy
  */
 export function generateRandomSecretKey(entropy?: Buffer): Buffer {
```

Comment:
This still seems off to me. These are the primitives that I think should be exposed somewhere:
- create master key from seed entropy
- create seed entropy from mnemonic
- derive a child private key given a private key and a path

we currently have a sort of high level convenience interface:
- create master key from seed entropy + random
- mnemonic to child key
- seed to child key

Reply:
- create master key from seed entropy - that happens if you don't pass path param
- create seed entropy from mnemonic - thats not bls related, thats scope of bip39 lib
- derive a child private key given a private key and a path - not sure the use case of this, I would skip this until someone requests this though

Reply:
I would prefer to see these as unique functions that don't do optionally different things. More functional.
eg:
```typescript
function mnemonicToSeed(mnemonic): seed;
function createPrivateKey(seed): key;
function deriveChildKey(key, path): key;
```
Now instead of having one "convenient" interface that does several things at once, the functions are separated, do one thing, and can be used in any configuration:
```typescript
function mnemonicToChildKey(mnemonic, path) {
  return deriveChildKey(createPrivateKey(mnemonicToSeed(mnemonic)), path);
}
```

Reply:
I don't get it. Functions are quite separated, every one of them generates key, if you submit path it will return child key otherwise it would just return regular key.

`mnemonicToSeed` is function that is not related to this library whatsoever, it's bip39. Exposing this would just consufe users as they won't be able to differ between valid private key and seed.

`createPrivateKey` and `deriveChildKey` would imply that you do `deriveChildKey(createPrivateKey(), path)` which is quite confusing as in should deriveChildKey convert key to valid master key or keep it as is, do we verify if given key is in valid order?

Reply:
Returning a master key is different behavior than returning a child key. Returning both in a single function is confusing.

We currently do a `deriveChildKey(createPrivateKey(), path)`, just without exposing the raw underlying functions.
```
  return pathToIndices(path).reduce(
    (parent, index) => {
      return deriveChildSK(parent, index);
    },
    masterKey
  );
```

I see what you're saying. You want to protect users from dealing with the underlying components. I'm talking about exposing the components of eip 2333 and eip 2334 so that they can be reused elsewhere. Not just assuming we know best and only exporting a high level interface. Maybe we should separately publish a bls-eip-2334?

Perhaps `deriveChildKey` isn't the right thing to expose, but at the very least, I think we should expose `deriveMasterSK` and `deriveChildSK`, and remove the optional part of the  `path` param in `mnemonicToSecretKey` and `deriveKey`

Reply:
> Returning a master key is different behavior than returning a child key. Returning both in a single function is confusing.

I disagree, master and child are secret key in it's core it's just different path ("m" vs "m/12381/60/0/0").

I'm fine with publishing bls-eip-2334 but I don't see much of a point in that library by itself...

Reply:
The difference there is that we're equating the `"m"` path with `null` (but with a different code path). This special case is added for no good reason and just complicates the code and interface.

The point of exposing the eip 2334 machinery by itself is that it lets others experiment with different ways of storing/traversing the key hierarchy.
Eg: if someone wants a more stateful representation of the key and path (like your original implementation), they might want to use `deriveMasterSK` and `deriveChildSK` directly.

Reply:
That equal to saying lets publish every method of lodestar separately because someone might experiment with it.

This library is suppose to cover two most common use cases "mnemonic to key" and "random to valid key". If people want's to experiment they are welcome to fork library and experiment though.

Reply:
Not really. Is it unthinkable that someone would want to crawl their hdwallet looking for balances? There are plenty of reasons someone may want the exposed hd wallet functions.

I'm not viewing this library only through the lens of our immediate needs, I'm thinking about what makes sense to expose as a general utility library.

> This library is suppose to cover two most common use cases "mnemonic to key" and "random to valid key". 

This is where my sticking point is. This hasn't been immediately clear from looking at the code in this library.
I'm been looking for a utility library that exposes `deriveMasterSK`, `deriveChildSK`, and `pathToIndices`, because these are the primitive building blocks.

Reply:
well you are welcome to create separate library with just utility functions(this library is called bls-keygen :smile: ), but I'm pretty sure this library contains everything you would possibly need. 

You can crawl hd wallet by just generating different paths though.

Reply:
Ok, I'll pull out the utility functions into a `bls-hd-key` that we can use here.